### ISSUE TITLE
 115 Illegal instruction

### ISSUE THREAD

**Issue by lishaojun616** [Mintplex-Labs/anything-llm#1331]
The anythingllm is installed in  Ubuntu server. 
In the system LLM set ,the system can connect to the Ollama server and get the models .
But when chat in workspace ,the docker is exited.
1.Show the info in browser:
 ![image](https://github.com/Mintplex-Labs/anything-llm/assets/135484168/3d987c50-d826-479f-9a8a-431638a99330)
2.and the docker logs:
"/usr/local/bin/docker-entrypoint.sh: line 7:   115 Illegal instruction     (core dumped) node /app/server/index.js"

What's the problem?

**Comment by timothycarambat** [comment]
Docker engine - it appears
https://github.com/Mintplex-Labs/anything-llm/issues/1290#issuecomment-2101960232

That issue is specifically occurring on Mac, but it is the same on Linux/Ubuntu as well.

**Comment by SyuanYo** [comment]
Hi @lishaojun616 , have you managed to resolve the issue? 
I am experiencing the same situation as you described in #1323 .

**Comment by joachimt-git** [comment]
Hello everyone,

I face the same problem. 

Same setup, ubuntu 22.04 LTS, using ollama as llm.

I have installed the newest docker engine, build anything-llm with docker-compose-v2.

```
...
[TELEMETRY SENT] {
  event: 'workspace_created',
  distinctId: 'c060354a-b171-4702-b83a-9da2ef0612e4',
  properties: {
    multiUserMode: false,
    LLMSelection: 'ollama',
    Embedder: 'native',
    VectorDbSelection: 'lancedb',
    runtime: 'docker'
  }
}
[Event Logged] - workspace_created
[TELEMETRY SENT] {
  event: 'onboarding_complete',
  distinctId: 'c060354a-b171-4702-b83a-9da2ef0612e4',
  properties: { runtime: 'docker' }
}
[NativeEmbedder] Initialized
/usr/local/bin/docker-entrypoint.sh: line 7:   119 Illegal instruction     (core dumped) node /app/server/index.js
gitlab-runner@gradio:~$ docker --version
Docker version 26.1.3, build b72abbb
```
This issue is marked as closed. Is there a solution available?

Best regrads 

Joachim

**Comment by joachimt-git** [comment]
Hi Timothy,

I think what @xsn-cloud and I have in commun is that we both use ollama.

May that ne the cause of the failure?

Joachim

**Comment by xsn-cloud** [comment]
@timothycarambat Thanks for addressing this issue. Please let me know if there's anything I can assist you with.

I've conducted the following experiment, also considering that it might be an issue with Docker running on VMs and to verify resource issues:

- Clean install of Debian 12 on **baremetal** - Dual  Xeon E5-2650 v2 @ 2.60GHz with 96GB of RAM
- Docker version 26.1.3, build b72abbb
- Followed the procedure **exactly** as you did in  your previous comment (the same one you used on an Ubuntu 22.04 LTS AWS) 
- No documents loaded
- During the onboarding, anything-llm successfully communicates with ollama, accurately retrieves the installed models, and allows the selection of the model without issues.

This is the outcome after the onboarding when attempting to send a "hello" in a new chat. (docker logs -f [containerid]). 

Please note that the container was restarted from scratch to ensure the clarity of the logs.


```
Collector hot directory and tmp storage wiped!
Document processor app listening on port 8888
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma

✔ Generated Prisma Client (v5.3.1) to ./node_modules/@prisma/client in 338ms

Start using Prisma Client in Node.js (See: https://pris.ly/d/client)

'''
import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient()
'''


or start using Prisma Client at the edge (See: https://pris.ly/d/accelerate)
'''
import { PrismaClient } from '@prisma/client/edge'
const prisma = new PrismaClient()
'''

See other ways of importing Prisma Client: http://pris.ly/d/importing-client

Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": SQLite database "anythingllm.db" at "file:../storage/anythingllm.db"

20 migrations found in prisma/migrations


No pending migrations to apply.
┌─────────────────────────────────────────────────────────┐
│  Update available 5.3.1 -> 5.14.0                       │
│  Run the following to update                            │
│    npm i --save-dev prisma@latest                       │
│    npm i @prisma/client@latest                          │
└─────────────────────────────────────────────────────────┘
[TELEMETRY ENABLED] Anonymous Telemetry enabled. Telemetry helps Mintplex Labs Inc improve AnythingLLM.
prisma:info Starting a sqlite pool with 33 connections.
fatal: not a git repository (or any of the parent directories): .git
getGitVersion Command failed: git rev-parse HEAD
fatal: not a git repository (or any of the parent directories): .git

[TELEMETRY SENT] {
  event: 'server_boot',
  distinctId: '4f39e3fb-ac8c-4043-9586-c21ef46b0c47',
  properties: { commit: '--', runtime: 'docker' }
}
[CommunicationKey] RSA key pair generated for signed payloads within AnythingLLM services.
Primary server in HTTP mode listening on port 3001
[NativeEmbedder] Initialized
/usr/local/bin/docker-entrypoint.sh: line 7:   163 Illegal instruction     (core dumped) node /app/server/index.js

```

Thanks a lot for your time.

**Comment by timothycarambat** [comment]
If you were to not use the native embedder, this problem would not surface. The only commonality between all of this is varying CPUs. Transformers.js which runs the native embedder, uses [ONNX runtime](https://onnxruntime.ai/) and at this point the root cause has to be coming from there as this only occurs when using the native embedder and that is the supporting libraries to enable that functionality.

**Comment by timothycarambat** [comment]
Does the CPU you swapped to support AVXv2?

**Comment by computersrmyfriends** [comment]
No, my CPU does not support AVX2 however it supports AVX

**Comment by jorgen-k** [comment]
> Does the CPU you swapped to support AVXv2?

I am sorry, i do not know how to check that, i just changed o "host" being a Intel Core i9-9900K CPU

**Comment by timothycarambat** [comment]
At this time, the working hypothesis is that since Transformers.js uses ONNX runtime it will fail to execute any model (including the built in embedder) if AVX2 is not supported
https://github.com/microsoft/onnxruntime

@jorgen-k https://www.intel.com/content/www/us/en/products/sku/186605/intel-core-i99900k-processor-16m-cache-up-to-5-00-ghz/specifications.html
> Instruction Set Extensions Intel® SSE4.1, Intel® SSE4.2, Intel® AVX2

Supports AVX2

**Comment by joachimt-git** [comment]
I am using kvm as hyperviser and the virtual cpu doesn't support avx2. When I configure a passthrough of the cpu (which supports avx2)  as @jorgen-k has suggested it works for me as well.

**Comment by computersrmyfriends** [comment]
I ran out of luck for the AVX cpu. It's a XEON 2660 but only supports AVX. Had to find another machine.

**Comment by Smocvin** [comment]
I had the same issue with "/usr/local/bin/docker-entrypoint.sh: line 7: 115 Illegal instruction (core dumped) node /app/server/index.js". It seems the new Docker images of AnythingLLM have some issues, possibly on older systems. To fix the issue, I tried using older Docker versions and previous AnythingLLM images. While the older Docker versions did not resolve the issue, the older AnythingLLM images worked great. The newest working version for me was "sha256:1d994f027b5519d4bc5e1299892e7d0be1405308f10d0350ecefc8e717d3154f". You can find it here: https://github.com/Mintplex-Labs/anything-llm/pkgs/container/anything-llm/209298508

**Comment by timothycarambat** [comment]
@Smocvin, excellent work. Okay, then that pretty much nails down commit ca63012c0f569ad775b6fd22b9b7965d61812d77 as the issue commit. In that commit we moved from lancedb `0.1.19` to `0.4.11` (which is what we use on desktop version).

However, given how this issue seems to only be a problem with certain CPUs we have two choices:
Bump to `0.5.0` and see if that fixes it or roll back to `0.1.19`. Given how we do not leverage or dive deep into LanceDBs API much, the code change is quite minimal or none.

What I will need though is some help from the community as I do not have a single machine, VM, or instance that I can replicate this bug with. So my ask is:

- If you are getting this bug, and it is on a Cloud-container service. What service and instance specs are you using so we can provision a test instance for replicate and debugging.

_or_

If anyone is willing to help debug the hard way I am going to creat two new tags on docker `:lancedb_bump` `:lancedb_revert` and I would need someone suffering from this issue to pull both and see which works. 

Obviously if we can bump up, that would be ideal, but I would rather not field this issue for the rest of time since lancedb should _just work_.

### Links to images
`lancedb_bump`: _TBD_
`lancedb_revert`: _TBD_

Will update this comment when those images exist.

**Comment by Dozer316** [comment]
I'm happy to help debug here locally with the newly created image tags when available.   I have two machines I can test on here with AVX (Debian docker) and AVX2 (Windows docker desktop).   I get the core dump on the AVX machine with :latest but the AVX2 machine runs the container fine so I can provide output from both of them if needed.

**Comment by timothycarambat** [comment]
@Dozer316 @acote88 @computersrmyfriends can any of you who have this issue on the `master/latest` image check and see if `lancedb_bump` or `lancedb_revert` work on the impact machine?

Hopefully the `_bump` image works, otherwise we are in for some pain, but at least we can debug from there. I am friends with the LanceDB team so I can escalate to them if the issue persists.

**Comment by Dozer316** [comment]
Hey there - revert has solved the problem on the impacted machine, bump still core dumps unfortunately. 

Thanks for taking a look at this for us.

**Comment by xsn-cloud** [comment]
lancedb_bump: Crashes
lancedb_db: Works

CPU: AVX only
Using an external embedder (ollama also)

**Comment by timothycarambat** [comment]
Thank you @Dozer316 @acote88 @xsn-cloud for all taking the time to test both, which is very tedious. Ill contact the lancedb team as well as see if we can rollback the docker `vectordb` package in the interim.

**Comment by cyberlink1** [comment]
I just closed my report out #1618 because it was caused by the same thing. AVX was not a flag on the virtual CPU.

**Comment by timothycarambat** [comment]
Okay, so the reason this issue occurs is due to LanceDB having its minimum target of `haswell` as of version ~0.20. This is because performance on AVX2 is just much better and haswell/AVX is really old.

So right now there are two options to go around this:
- Upgrade or migrate use to a CPU that supports AVX2
- We can maintain an image that works with the older vectordb package. Truthfully, I really dont want to do that and we have to draw the line somewhere. The reason I don't is that while code-change is minor this likely will become increasingly burdensome to maintain as we continue to bump lance into later versions. Knowing where the issue lies though is very useful.

Either way, the root cause is the requirement of the underlying CPU to have AVX2. Closing currently as `wontfix` but discussion is still open for any more commentary.

**Comment by Nododot** [comment]
You have no idea how long I've had to search everywhere and how many reconfigurations and reinstalls I did before I found this thread. Could you MAYBE write SOMEWHERE that currently AnythingLLM requires an AVX2 CPU to work properly?

**Comment by some-random-sperg** [comment]
Hello @timothycarambat 

Thank you for publishing the [Lancedb_revert image](https://hub.docker.com/layers/mintplexlabs/anythingllm/lancedb_revert/images/sha256-6fe0d0eb1ce029b65e4e231457f1c0029f809bc74314e942bd5aff84bdff598c?context=explore) at all in the first place. 

Currently, googling the error message took me to [this thread ](https://github.com/Mintplex-Labs/anything-llm/issues/1651), which in turn links to this one. 

To resolve the issue all I had to do was update the docker run command with the `lancedb_revert` tag, and otherwise "off it went" 

My pc is old, but it's what I've got and sadly upgrading just isn't on the cards any time soon - I'm grateful to have a way to try it out at all.

I appreciate it's unreasonable to put in ongoing effort for small subset of users running into incompatibility problems because they insist on using a relic from the before times - Especially since it's going to start increasingly cropping up elsewhere as well. 

Having an image an in the first place is great, but it'd be nice if there was some way to "run out the clock" on updates until breaking changes inevitably came along. 

Would it be possible to have an unsupported update that pins the version of lancedb in place, dumps latest and/or dev over the top of it and "When it breaks, that's the end of the ride.... May the odds be ever in your favour"?

When it does, ideally the docker image gets a 2nd "unsupported final build" release based on that point version and that's the end of that.

**Comment by kaomoneus** [comment]
Just bumped into the same issue. I'm running anythingllm under vps. Afaik it doesn't even supports avx.
TLDR: guys who have this issue, just pull `mintplexlabs/anythingllm:lancedb_revert` instead of latest tag.

Nevertheless, I think there should be some gracefull way to process platform incompatibilities..

**Comment by Gamechiefx** [comment]
THANK YOU.... THANK YOU... FOR MENTIONING THE AVX2 SUPPORT

Where is this documented?

**Comment by timothycarambat** [comment]
https://docs.anythingllm.com/fetch-failed-on-upload#your-cpu-is-not-supported

I should update the link since this is for `fetch failed` on uploading, but looks like it can also arise from this error message as well
